crate/findex/src/batcher_findex.rs

```diff
@@ -0,0 +1,299 @@
+use std::{collections::HashSet, fmt::Debug, hash::Hash, sync::Arc};
+
+use cosmian_sse_memories::{ADDRESS_LENGTH, Address, BatchingMemoryADT, MemoryBatcher};
+
+use crate::{Decoder, Encoder, Findex, IndexADT, adt::IndexBatcher, error::BatchFindexError};
+
+#[derive(Debug)]
+pub struct FindexBatcher<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    EncodingError: Send + Debug,
+    BatcherMemory: Clone + Send + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> {
+    memory: BatcherMemory,
+    encode: Encoder<Value, BatcherMemory::Word, EncodingError>,
+    decode: Decoder<Value, BatcherMemory::Word, EncodingError>,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    BatcherMemory: Debug
+        + Send
+        + Sync
+        + Clone
+        + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+    EncodingError: Send + Debug,
+> FindexBatcher<WORD_LENGTH, Value, EncodingError, BatcherMemory>
+{
+    pub fn new(
+        memory: BatcherMemory,
+        encode: Encoder<Value, BatcherMemory::Word, EncodingError>,
+        decode: Decoder<Value, BatcherMemory::Word, EncodingError>,
+    ) -> Self {
+        Self {
+            memory,
+            encode,
+            decode,
+        }
+    }
+
+    // Both insert and delete operations make an unbounded number of calls to
+    // `guarded_write` on the memory layer.
+    async fn batch_insert_or_delete<Keyword>(
+        &self,
+        entries: Vec<(Keyword, impl Send + IntoIterator<Item = Value>)>,
+        is_insert: bool,
+    ) -> Result<(), BatchFindexError<BatcherMemory>>
+    where
+        Keyword: Send + Sync + Hash + Eq,
+    {
+        let mut futures = Vec::new();
+        let memory = Arc::new(MemoryBatcher::new(self.memory.clone(), entries.len()));
```

Comment:
It may be clearer if the memory batcher API requires us to make an explicit call to some subscription function to have more parallelism with the call to `unsubscribe`: `memory.subscribe(entries.len())`. Doing so may have drawbacks and is therefore open for discussion.